TiKV Command Line RFC

Hoverbear opened the pull request:
Based on #7 we discussed having a command line client.

This RFC proposes one! How exciting.

Hoverbear commented:
@sunxiaoguang can you take a look since it's based off your RFC? :)

sunxiaoguang commented:
> @sunxiaoguang can you take a look since it's based off your RFC? :)

Sure. I will have a look in a moment.

sunxiaoguang commented on text/2019-02-26-tikv-client-command-line.md:
````diff
@@ -0,0 +1,208 @@
+```
+
+One-off transactional mode is a bit different, as we need some way to maintain
+timestamp across commands. The user may choose how to do this, the reccomended
````
recommended

Hoverbear replied:
Oh, damn. Thanks.

sunxiaoguang commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,208 @@
+
+## Drawbacks
+
+We may not wish to need to maintain this tool, or we may wish for the community
```
'wish to' or 'need to' ?

Hoverbear replied:
I think I can reword this, but I meant to say "It increases maintenance burden"...

sunxiaoguang commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,208 @@
+  call.
+* (*Use `tikv-cli`*), we could use `tikv-cli` for this, but it creates as fuzzy
+  distinction between management tools and clients.
+* (*No REPL, or only REPL*). We may chose to not have a REPL mode, or we may chose
```
I rarely use one off mode of mysql client. But I think running commands from file is very useful.

Hoverbear replied:
Is there an advantage to just piping to the REPL?

sunxiaoguang replied:
Hm, it actually doesn't make much difference. It's just something that I would expect to have as it is in mysql and hive client.

Hoverbear replied:
What do you think @siddontang ?

nrc replied:
I think it is good to have both, and I don't think it adds too much complexity (as long as we can remove the need for raw and txn modes as described above).

nrc commented on text/2019-02-26-tikv-client-command-line.md:
````diff
@@ -0,0 +1,208 @@
+minify = false
+# Level of logging to use. Available: error. warn, info, debug, trace
+logging = "info"
+```
````
What are the values here? I expected them to be defaults, but that doesn't match the text. (I would also expect the default log level to be `error`, or possibly `warning`.

Hoverbear replied:
Should be all of these: https://docs.rs/slog/2.4.1/slog/enum.Level.html, good catch.

nrc commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,208 @@
+JSON was chosen because it is an efficient, pervasive, human-readable, modern
+standard that has many tools (eg `jq`) to use for post processing if needed.
+
+All standard (non-logging) output will be on `stdout`.
```
To clarify, "standard (non-logging) output" means the json stuff above? It would be good to mention that the total output on stdout will be valid JSON (will this be the case for stderr too?).

Hoverbear replied:
Yes the intention is to make sure the output is fairly consistent on stdout so folks can use it in scripts.

nrc commented on text/2019-02-26-tikv-client-command-line.md:
````diff
@@ -0,0 +1,208 @@
+    { key: "tikv", value: "rust" },
+]
+Finished in 0.003s.
+```
````
These "Finished in 0.003s.", will they be on stderr or stdout? Are they configurable?

Hoverbear replied:
I think these should be stderr.

Hoverbear replied:
Added a switch for these

nrc commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,208 @@
+Users will interact with the tool through the command line. It will support both
+a one-off mode or a REPL (Read-eval-print-loop) style mode (See alternatives).
+
+Example of the one-off raw mode:
```
Could you explain what "raw" means before using it here.

nrc commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,208 @@
+
+REPL transactional mode, notice how the `>>` communicates that the user is in a
+transaction. Also notice how the transaction timestamp is outputted in the
+request complete messages.
```
Why do we need a transactional one-off mode? Could we take multiple commands as input and thus avoid making a distinction between raw and txn modes?

Hoverbear replied:
In theory transactional one-off won't be used much.

nrc commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,208 @@
+## Drawbacks
+
+We may not wish to need to maintain this tool, or we may wish for the community
+to maintain it.
```
I think the drawback here is the maintenance burden. We'll need to find some way to handle that. I don't think we need to specify how we handle that here, just be aware that there is this cost.

Hoverbear replied:
Does the sentence present not communicate this correctly? Because that's exactly what I meant to say. :)

nrc commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,208 @@
+  to *only* have a REPL mode.
+* (*A different timestamp handling for one-off mode*). We may find the way we
+  handle timestamps in transactional one-off mode to be unwieldly. Perhaps there
+  is a better way.
```
Is this standard in other clients? (MySQL in particular). If not, I would think some way of handling multiple commands at once would work. I guess the use case we want to handle is a script which adds commands to a transaction based on the results of earlier commands in the transaction. Unless there is some clear need for that, then I'd imagine that use case is better supported using an API and we should explicitly not support it.

Hoverbear replied:
Part of the reason it's defined this way is that we don't have a text representation (Eg `BEGIN SELECT * FROM foo; ELECT * FROM bar; COMMIT`)

The only other idea I had to support this use case is `tikv-client txn set foo bar then get bar` or something like that and use a seperator... But I'm not sure this is nicer.

Hoverbear commented:
After some discussion with @nrc we think that it is appropriate to default to `transaction` in most cases. So perhaps we should default to `transaction` and allow the user to configure it via the config with `--mode`

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+## Summary
+
+Use the TiKV Rust Client to build a simple command line tool for inspection and anaylsis.
```
```suggestion
This article describes how to the TiKV Rust Client to build a simple command line tool for inspection and analysis.
```

Hoverbear replied:
This is an RFC, so this is a proposed action, not an instruction guide for someone. :) Thanks for the analysis correction!

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+## Motivation
+
+Currently `tikv-ctl` offers some ways to interact with a TiKV cluster, it
```
```suggestion
Currently `tikv-ctl` offers some ways to interact with a TiKV cluster, which
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+focuses mostly on the *management* of a TiKV cluster instead of a *client* for
+the cluster.
+
+With the `tikv-client` Rust crate users will find a way to interact with their TiKV
```
```suggestion
With the `tikv-client` Rust crate, users will find a way to interact with their TiKV
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+application.
+
+Building a simple command line tool to execute commands such as `get`, `set`,
+`delete`, and `scan` would mean users do not need to worry about creating such
```
```suggestion
`delete`, or `scan` means users do not need to worry about creating such
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+It also provides us with a useful tool during the development cycle, as it will
+be very easy to issue one-off commands to TiKV.
+
+Further, we mentioned in the Rust Client RFC
```
```suggestion
Plus, this tool is one of the goals we mentioned in the Rust Client RFC
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+be very easy to issue one-off commands to TiKV.
+
+Further, we mentioned in the Rust Client RFC
+(https://github.com/tikv/rfcs/pull/7) we mentioned we wanted to accomplish this.
```
```suggestion
(https://github.com/tikv/rfcs/pull/7).
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+## Detailed design
+
+The client will be created using the `tikv-client` crate and its already existing
```
```suggestion
The client will be created using the `tikv-client` crate and existing
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+## Detailed design
+
+The client will be created using the `tikv-client` crate and its already existing
+functionality. It will utilize `clap` to provide a simple wrapper around the
```
```suggestion
functionality of the crate.  `clap` will be used to provide a simple wrapper around the
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+### Install
+
+We will build this tool in the `tikv-cli` repository. It will be published.
```
```suggestion
This tool will be built in the `tikv-cli` repository and published.
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+### Configuration
+
+The tool will retrieve its configuration from one of the following places, in
+order of precedence (greatest to least):
```
```suggestion
an order of precedence (greatest to least):
```

Hoverbear replied:
"in an order of precedence" sounds like there are several orders instead of just one.

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+The tool will retrieve its configuration from one of the following places, in
+order of precedence (greatest to least):
+
+* The command line options, such as `--pd 127.0.0.1:2379`.
```
```suggestion
* Command line options, such as `--pd 127.0.0.1:2379`.
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+* The command line options, such as `--pd 127.0.0.1:2379`.
+* A local configuration file `./.tikv-cli.toml`
+* The global configuration file `${XDG_CONFIG_HOME}/tikv-cli.toml`.
```
```suggestion
* A global configuration file `${XDG_CONFIG_HOME}/tikv-cli.toml`.
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+* The command line options, such as `--pd 127.0.0.1:2379`.
+* A local configuration file `./.tikv-cli.toml`
+* The global configuration file `${XDG_CONFIG_HOME}/tikv-cli.toml`.
+* The default options (connect to localhost, no TLS).
```
```suggestion
* Default options (connecting to localhost, no TLS).
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+### Input/Output
+
+By default input is by positional arguments in `bash`/`sh` string format.
```
by positional arguments？

Hoverbear replied:
Yes, eg

```bash
tikv-cli get foo
         ^--^--- Positional Arguments
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+array. Errors will use an `{ error: "" }` object. Opting out of prettyprint will
+be via the `--minify` flag.
+
+JSON was chosen because it is an efficient, pervasive, human-readable, modern
```
```suggestion
JSON is chosen because it is an efficient, pervasive, human-readable, modern
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+### Interacting with it
+
+Users will interact with the tool through the command line. It will support both
```
```suggestion
Users will interact with the tool through command line. It will support both
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+### Interacting with it
+
+Users will interact with the tool through the command line. It will support both
+a one-off mode or a REPL (Read-eval-print-loop) style mode (See alternatives).
```
Since it's the mode name, "a" is not needed here.
```suggestion
one-off mode and REPL (Read-eval-print-loop) style mode (See [Alternatives](#Alternatives)).
```

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+Users will interact with the tool through the command line. It will support both
+a one-off mode or a REPL (Read-eval-print-loop) style mode (See alternatives).
+
+TiKV has two APIs, *raw* and *transactional*, both are supported. Learn about the
```
"both are supported" where? In this tool?  Seems missing something.

dcalvin commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+It will feature configurable logging levels with `slog` which will output to `stderr`.
+
+The tool will exit zero on successful requests, and non-zero on unsuccessful
```
Can you explain exit zero to me?

Hoverbear replied:
When a program finishes it returns an exit code, if it is zero it is successful, if it is not zero it is generally not successful. https://en.wikipedia.org/wiki/Exit_status

I clarified it.

Hoverbear commented:
@sunxiaoguang Do you think you could lend your eyes again? It should be ready.

AndreMouche commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+
+By default input is by positional arguments in `bash`/`sh` string format.
+
+By default the tool will output in prettyprinted JSON. For single entry results,
```
```suggestion
By default, the tool will output in pretty printed JSON. For single entry results,
```

AndreMouche commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,232 @@
+]
+Finished in 0.003s. (txn ts: 1551216224838)
+>> commit
+Finished in 0.001s. (txn ts: 1551216224838)
```
Could we print the `commit_ts` too?

Hoverbear replied:
Always or configurable?

AndreMouche replied:
both are ok for me ^_^

AndreMouche commented on text/2019-02-26-tikv-client-command-line.md:
````diff
@@ -0,0 +1,232 @@
+Finished in 0.001s. (txn ts: 1551216224838)
+```
+
+One-off transactional mode is a bit different, a series of commands can be
````
I think `Auto-commit mode` will be more understandable for me ^^

Hoverbear replied:
I clarified the description and added that term :)

Hoverbear commented:
@AndreMouche We have that at https://github.com/tikv/client-rust already. :)

brson commented:
Sweet idea, and fun.

A repl requires building a language and interpreter and there will probably be ongoing feature requests to expand its capabilities. It might be easier to create python bindings and reuse python's repl. Or maybe some other scripting language that's easy on the eyes.

Hoverbear commented:
😵 Then it's annoying to install and build. It's much easier to distribute a single musl binary or allow a simple `cargo install` instead of having to mess around with system packages and pip.

I suspect it won't be terrifically hard to create a little repl with clap's `arg_matches` and a loop. Might try to whip up a demo...

Hoverbear commented:
@brson Here's a simple REPL example for interactive mode. It'll parse first token (blah, "blah blah", 123) as the KEY, optionallly the second as the VALUE using the normal clap parser and a simple lexer and readline interface:

```rust
use clap::{App, Arg};
use std::env::args_os;
use linefeed::{Interface, ReadResult};
use scanlex::{Scanner, Token};

fn main() {
    let repl_interface = Interface::new("tikv").unwrap();
    repl_interface.set_prompt("tikv > ").unwrap();

    let mut app = App::new("get")
        .arg(Arg::with_name("KEY")
            .required(true))
        .arg(Arg::with_name("VALUE")
            .required(false));

    let matches = app.get_matches_from_safe_borrow(args_os());
    println!("{:?}", matches);

    loop {
        if let Some(res) = repl_interface.read_line_step(None).unwrap() {
            match res {
                ReadResult::Input(input) => {
                    let scanner = Scanner::new(&input);

                    let tokens = scanner.map(|v| match v {
                        Token::Iden(x) | Token::Str(x) => x,
                        Token::Int(y) => format!("{}", y),
                        _ => panic!("No arg"),
                    });

                    let preface = vec![String::from("tikv")];
                    let iter = preface.into_iter().chain(tokens.into_iter());

                    let matches = app.get_matches_from_safe_borrow(iter);
                    println!("{:?}", matches);
                }
                ReadResult::Eof => {
                    break;
                },
                _ => {}
            }
        }
    }
    println!("{:?}", matches);
}
```

You can have a try and see that it feels fine to use and doesn't require a lot of work.

brson commented:
> Then it's annoying to install and build. It's much easier to distribute a single musl binary or allow a simple `cargo install` instead of having to mess around with system packages and pip. Besides! Python bindings are a whole other project.

That's a problem with Python specifically, yes. There may be other suitable languages.

ice1000 commented:
For REPL I'll recommend rustyline which can do completion.

Hoverbear commented:
Hi @sciamp-dev, thanks for volunteering to get involved with this project! We're really excited to learn with you and I'm looking forward to working with you directly.

I know you've already had a look at this RFC, but could you please verify it and give your feedback/approval?

@zhangjinpeng1987 @brson do you think you could take a look at this project and give feedback so our friend can get started? @siddontang do you think you could help me open a `client-cli` repo for this and give me maintainer status? I'll work with @sciamp-dev to create an initial implementation in the coming weeks.

sciamp-dev commented:
Hi everyone, I'm exited to work with you, too! And thanks @Hoverbear :)

I agree with the use of RustyLine for REPL mode, reading the docs it looks more complete and it has more functionalities then other crates like liner or linefeed.
As for the idea of making Python bindings, if there's a solution in Rust I think it's better and more user-friendly.

ice1000 commented:
I'd say making use of some existing Rust-implemented scripting languages and add binding for them or create our own PL are both good ideas.

brson commented:
I think somebody should go ahead and get started &mdash; I'd rather have code written than not. Whatever that tool looks like will inevitably evolve and undoubtedly end up different. I would not be surprised if the first iteration ends up scrapped, redesigned and rewritten with lessons learned.

So that's what I think should happen, but below are some things to consider.

DSLs have a mixed track record. I'm not though going to seriously advocate binding a scripting language for the DSL. The examples in this RFC are very simple, and if that ends up being the extent of the tool's functionality then a DSL is fine. A sprawling DSL is a poor programming language.

The duality between the one-off mode and the repl probably imposes yet-unknown limits on the grammar.
The one-off mode accepts the tokens of its grammar as individual command line arguments, which has limitations compared to the arbitrary string parsing possible in the repl. Other scripting languages tend to have a flag that just accepts a single string containing script text. With the simple grammar implied here, just statements, keywords, and strings, no variables, no expressions, the duality between cli arguments and tokens looks neat. But that design might not hold up.

A limited language for the one-off mode can be augmented with other tools like `jq`, but that isn't true for the repl. The repl described is not powerful.

Hoverbear commented:
🎉 @sciamp-dev please review and give feedback/approve this PR, I've reached out to the benevolent rulers of the TiKV org to create the repo for us. :)

sciamp-dev commented:
Sure @Hoverbear ! I approve this PR, we can go ahead and start.
I agree with what @brson said. I'm doing some quick tests with RustyLine, PyO3(/cpython) and wasm-bindgen for binding with Python and JS. But I guess we'll choose/adapt considering whatever needs we'll have.

Hoverbear commented:
We (@siddontang , @sciamp-dev ) discussed this on our community chat and decided to start with this being part of the `client-rust` repo and making it an opt-in binary until the project fully matures. This will help it act as a nice test bed for the Rust client, and give it some extra maintenance eyes.

I also corrected some mistakes around `tikv-cli` vs `tikv-client`.

I also added a note about output formats as we do have some commonly used formats and we need to consider our non-UTF-8 users. Please, take a look! :)

sciamp-dev commented on text/2019-02-26-tikv-client-command-line.md:
````diff
@@ -0,0 +1,256 @@
+```toml
+[connection]
+# A list of `pd` endpoints.
+# Default: None
````
As for line 51 of this RFC, default should be localhost without TLS

```suggestion
# Default: 127.0.0.1:2379 (no TLS)
```

sciamp-dev commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,256 @@
+# TiKV has a Unified Key Format which is used in cases where a more human readable format is
+# required for non-UTF-8 data.
+# Avaiable: utf-8, ukf, profobuffer, hex
+key-encoding = "utf-8"
```
```suggestion
key_encoding = "utf-8"
```

sciamp-dev commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,256 @@
+# required for non-UTF-8 data.
+# Avaiable: utf-8, ukf, profobuffer, hex
+key-encoding = "utf-8"
+value-encoding = "utf-8"
```
```suggestion
value_encoding = "utf-8"
```

sciamp-dev commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,256 @@
+
+TiKV is often used in cases where a key or value is not a UTF-8 compatible
+string, or may print undesirable control characters to the console. We provide the
+`key-encoding` and `value-encoding` configuration options to control this.
```
```suggestion
`key_encoding` and `value_encoding` configuration options to control this.
```

sciamp-dev commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,256 @@
+strings for both keys and values. In order to prevent confusion, we default to
+UTF-8 for both and make the options for `*-encoding` discoverable.
+
+The `key-encoding` and `value-encoding` options support `utf-8`, `ukf` (TiKV's
```
```suggestion
The `key_encoding` and `value_encoding` options support `utf-8`, `ukf` (TiKV's
```

sciamp-dev commented on text/2019-02-26-tikv-client-command-line.md:
```diff
@@ -0,0 +1,256 @@
+# Control the encoding.
+# TiKV has a Unified Key Format which is used in cases where a more human readable format is
+# required for non-UTF-8 data.
+# Avaiable: utf-8, ukf, profobuffer, hex
```
```suggestion
# Avaiable: utf-8, ukf, protobuffer, hex
```

Hoverbear commented:
Closing. I am not longer involved with the project and have no stake in this merging. Feel free to create a new PR.
